File: leetcode/scala-3/src/main/scala/com/leetcode/problems/medium/ContainerWithMostWater.scala
package com.leetcode.problems.medium

import scala.annotation.tailrec

/** LeetCode #11: Container With Most Water
  *
  *   - https://leetcode.com/problems/container-with-most-water/
  *   - https://www.hellointerview.com/learn/code/two-pointers/container-with-most-water
  */
object ContainerWithMostWater:
  def maxArea(hs: Array[Int]): Int =
    def area(l: Int, r: Int): Int = (r - l) * (hs(r) min hs(l))

    @tailrec
    def go(a: Option[Int] = None)(l: Int, r: Int): Option[Int] =
      a match
        case _ if l >= r                => a
        case Some(a) if a >= area(l, r) =>
          val (i, j) = if hs(l) < hs(r) then (l + 1, r) else (l, r - 1)
          go(Some(a))(i, j)
        case _ => go(Some(area(l, r)))(l, r)

    go(None)(0, hs.length - 1) match
      case Some(area) => area
      case _          => 0

File: hellointerview/src/two_pointers/container_with_most_water.rs
// https://www.hellointerview.com/learn/code/two-pointers/container-with-most-water

pub fn max_area(hs: Vec<i32>) -> i32 {
    let mut a = 0;

    let mut l = 0;
    let mut r = hs.len() - 1;

    while l < r {
        a = a.max((r - l) as i32 * hs[l].min(hs[r]));

        if hs[l] < hs[r] {
            l += 1
        } else {
            r -= 1
        };
    }

    a
}

#[cfg(test)]
mod tests {
    use super::max_area;

    #[test]
    fn two_equal_walls() {
        let heights = vec![1, 1];
        assert_eq!(max_area(heights), 1);
    }

    #[test]
    fn mixed_heights() {
        let heights = vec![1, 8, 6, 2, 5, 4, 8, 3, 7];
        assert_eq!(max_area(heights), 49);
    }

    #[test]
    fn zero_heights() {
        let heights = vec![0, 0];
        assert_eq!(max_area(heights), 0);
    }

    #[test]
    fn increasing_heights() {
        let heights = vec![1, 2, 3, 4, 5];
        assert_eq!(max_area(heights), 6);
    }

    #[test]
    fn decreasing_heights() {
        let heights = vec![5, 4, 3, 2, 1];
        assert_eq!(max_area(heights), 6);
    }

    #[test]
    fn maximum_length_and_height() {
        let heights = vec![10_000; 100_000];
        assert_eq!(max_area(heights), 999_990_000);
    }
}
